NDTensors/src/lib/TypeParameterAccessors/src/interface.jl

```diff
@@ -0,0 +1,12 @@
+### These are optional interface functions which can be 
+### defined on your type to make functions like `set_eltype` 
+### usable
+
+# Required overloads, generic fallback
+position(::Type, ::Function) = UndefinedPosition()
+
+function default_parameter(type::Type, fun::Function)
+  return error("The default parameter of function $(fun) is not defined for type $(type)")
+end
+
+default_parameters(::Type)::Tuple = ()
```

Comment:
Remove this in favor of `parameter_function`.

Reply:
Should this be `parameter_functions`?

Reply:
No, the idea would be to move to specifying one at a time, i.e. `parameter_function(::Type, ::Position) -> Function`, instead of defining them all at once in a `Tuple`. Of course there could be a wrapper around that called `parameter_functions(::Type)` which calls those functions in a loop over `nparameters(::Type)` positions and collects them into a Tuple.

Reply:
Okay this makes sense and I have developed a generic version which works and is type stable 
```
Base.@assume_effects :foldable function default_parameters(type::Type)
  return Tuple(map(i -> parameter_function(type, Position(i)), 1:nparameters(type)))
end
```

Reply:
Here's an interesting alternative implementation:
```julia
# Julia internals
@generated to_datatype(::Type{type}) where {type} = Base.unwrap_unionall(type)
parameters(type::DataType) = type.parameters
# Derived from Julia internals
to_datatype(type::DataType) = type
parameters(type::UnionAll) = parameters(to_datatype(type))
nparameters(type::Type) = length(parameters(type))
struct Position{Pos} end
Position(pos) = Position{pos}()
eachposition(type::Type) = ntuple(i -> Position(i), nparameters(type))
parameter_functions(type::Type) = map(pos -> parameter_function(type, pos), eachposition(type))

# Implement for `Array`.
parameter_function(::Type{<:Array}, ::Position{1}) = eltype
parameter_function(::Type{<:Array}, ::Position{2}) = ndims

# Test type stability
@code_warntype parameter_functions(Array)
@code_warntype parameter_functions(Array{Float64})
@code_warntype parameter_functions(Array{Float64,2})
```
I was having trouble making it all type stable, but it seems like defining `to_datatype` as a generated function is enough to make all the rest of it type stable.

Reply:
Also, maybe we should call it `parameter_name` instead of `parameter_function`, since in principle it doesn't need to be a function.

Reply:
This makes sense and I like the name change.  I was able to implement this and its type stable in all of my tests so far.